### PR TITLE
Zgc/diopi ascend use op plugin1

### PR DESCRIPTION
**Pull request by zhaoguochun1995**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Description

add  ascend op-plugin op to diopi ascend impl

## Checklist
**Before PR**:

- [ ] I have read and followed the workflow indicated in the [Contributors.md](https://github.com/DeepLink-org/DIOPI/blob/main/Contributors.md) to create this PR.
- [ ] Pre-commit or linting tools indicated in [Contributors.md](https://github.com/DeepLink-org/DIOPI/blob/main/Contributors.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: impl/ascend_npu/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,885 @@
...
+
+if(DEFINED ENV{ASCEND_CUSTOM_PATH})
+  set(ASCEND_DIR $ENV{ASCEND_CUSTOM_PATH})
+else()
+  set(ASCEND_DIR /usr/local/Ascend)
+endif()
+
+execute_process(
+    COMMAND sh -c "bash  ${CMAKE_CURRENT_SOURCE_DIR}/torch_npu/third_party/acl/libs/build_stub.sh"
+    COMMAND sh -c "echo 'download op-plugin ...'; cd ${CMAKE_CURRENT_SOURCE_DIR}/; git clone  https://gitee.com/ascend/op-plugin/ op-plugin || cd op-plugin && git checkout 186f4b87e2db4791a6ed697cea78d58a0ae92db1 && bash  ./gencode.sh 2.0 python "
```
Comment (yangbofun) [inline impl/ascend_npu/CMakeLists.txt:14]: 这个第三方库op plugin设为 impl/ascend下的third party?

File: impl/ascend_npu/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,885 @@
+cmake_minimum_required(VERSION 3.4)
+project(ascend_impl)
+
+set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} ${CMAKE_CURRENT_SOURCE_DIR})
+
+if(DEFINED ENV{ASCEND_CUSTOM_PATH})
```
Comment (yangbofun) [inline impl/ascend_npu/CMakeLists.txt:6]: 这里写为ASCEND_PATH吧

File: impl/ascend_npu/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,885 @@
...
+if(NOT DEFINED DIPU_ABI_V)
+  execute_process(
+    COMMAND
+      sh -x -c
+      "python -c 'import torch, builtins; print(next(item[-4:-2] for item in dir(builtins)      \
+          if \"__pybind11_internals_v4_gcc_libstdcpp_cxxabi10\" in item))'"
+    OUTPUT_VARIABLE DIPU_ABI_V)
+endif()
+
+if(NOT DEFINED DIPU_COMPILED_WITH_CXX11_ABI)
```
Comment (yangbofun) [inline impl/ascend_npu/CMakeLists.txt:52]: diopi应该是不知道dipu的存在的，diopi是c接口对外对接，这里是不是不用设置abi版本

File: impl/ascend_npu/diopi_impl/helper.hpp [hunk 4]
```diff
@@ -0,0 +1,450 @@
...
+
+
+#define PRIVATE_MACRO_VAR_ARGS_IMPL_COUNT(_1,_2,_3,_4,_5,_6,_7,_8,_9,_10,_11,_12,_13,_14,N,...) N
+#define PRIVATE_MACRO_VAR_ARGS_IMPL(args) PRIVATE_MACRO_VAR_ARGS_IMPL_COUNT args
+
+#define COUNT_MACRO_VARR(...) PRIVATE_MACRO_VAR_ARGS_IMPL((__VA_ARGS__,14,13,12,11,10,9,8,7,6,5,4,3,2,1))
+
+#define PRIVATE_CONCAT_STR2(x, y) x##y
+#define PRIVATE_CONCAT_STR1(x, y) PRIVATE_CONCAT_STR2(x, y)
+#define PRIVATE_CONCAT_STR(x, y) PRIVATE_CONCAT_STR1(x, y)
```
Comment (yangbofun) [inline impl/ascend_npu/diopi_impl/helper.hpp:112]: 这里只定义PRIVATE_CONCAT_STR就行了，1和2是不是多余了

File: impl/ascend_npu/diopi_impl/helper.hpp [hunk 5]
```diff
@@ -0,0 +1,450 @@
...
+
+}  // namespace cuda
+}  // namespace c10
+#endif
+
+namespace impl {
+
+namespace aten {
+
+inline void setCurCtx(diopiContextHandle_t ctx) {
```
Comment (yangbofun) [inline impl/ascend_npu/diopi_impl/helper.hpp:193]: 这里的setCurCtx和unsetCurCtx，写成一个函数getStream(diopiContextHandle_t ctx)返回昇腾的stream。

File: impl/ascend_npu/diopi_impl/helper.hpp [hunk 6]
```diff
@@ -0,0 +1,450 @@
...
+}
+
+inline const at::Tensor buildATen(diopiConstTensorHandle_t tensor) {
+    if (tensor == nullptr) return at::Tensor();
+    return *reinterpret_cast<const at::Tensor*>(tensor);
+}
+
+inline bool isInt(const diopiScalar_t* scalar) { return scalar->stype <= 7; }
+
+inline bool isFloat(const diopiScalar_t* scalar) { return scalar->stype > 7; }
```
Comment (yangbofun) [inline impl/ascend_npu/diopi_impl/helper.hpp:321]: 有个bool是>7的。

File: impl/ascend_npu/diopi_impl/helper.hpp [hunk 7]
```diff
@@ -0,0 +1,450 @@
...
+inline c10::OptionalIntArrayRef buildATen(diopiSize_t* size_ptr) {
+    if (size_ptr) {
+        return buildATen(*size_ptr);
+    }
+    return c10::nullopt;
+}
+
+
+template <typename T>
+inline decltype(auto) buildATenList(T* tensors, int64_t numTensors) {
```
Comment (yangbofun) [inline impl/ascend_npu/diopi_impl/helper.hpp:350]: 返回类型写成这个吧std::vector<at::Tensor> ，比较直观。

File: impl/ascend_npu/diopi_impl/helper.hpp [hunk 8]
```diff
@@ -0,0 +1,450 @@
...
+    auto at_##x4 = impl::aten::buildATen(x4); \
+    auto at_##x5 = impl::aten::buildATen(x5); \
+    auto at_##x6 = impl::aten::buildATen(x6); \
+    auto at_##x7 = impl::aten::buildATen(x7); \
+    auto at_##x8 = impl::aten::buildATen(x8); \
+    auto at_##x9 = impl::aten::buildATen(x9); \
+    auto at_##x10 = impl::aten::buildATen(x10);
+
+
+#define PRIVATE_MACRO_VAR_ARGS_IMPL_COUNT(_1,_2,_3,_4,_5,_6,_7,_8,_9,_10,_11,_12,_13,_14,N,...) N
```
[the remark this answers is not in the document]
Comment (zhaoguochun1995) [reply impl/ascend_npu/diopi_impl/helper.hpp:105]: 这个是内部使用的，写的短，如果有符号刚好重名了，会直接被替换掉导致出错，改短有啥意义呢

File: impl/ascend_npu/diopi_impl/helper.hpp (hunk 5 above)
Comment (zhaoguochun1995) [reply impl/ascend_npu/diopi_impl/helper.hpp:193]: 这里写这个是为了对齐op-plugin。op-plugin内部有获取流的接口。我们需要实现特定接口。getStream(diopiContextHandle_t ctx) 有类似的接口。先看看op-plugin

**Pull request merged**